### ISSUE TITLE
CPU appears to be stuck on initial frequency

### ISSUE THREAD

**Issue by krystian-hebel** [Dasharo/dasharo-issues#35]
**Dasharo version**
v0.4.0

**Dasharo variant**
Trustworthy Computing

**Affected component(s) or functionality**
Core frequency

**Brief summary**
CPU appears to be stuck on initial frequency

**How reproducible**
100%

**How to reproduce**

1. Boot into OS
2. `cat /proc/cpuinfo | grep clock`

**Expected behavior**
Clock frequency should be dynamically lowered to conserve power when idle, frequency may be different for different cores.

**Actual behavior**
All cores are stuck on the same, nominal frequency.

**Screenshots**
```
debian@debian:~$ cat /proc/cpuinfo | grep clock
clock		: 2700.000000MHz
clock		: 2700.000000MHz
clock		: 2700.000000MHz
clock		: 2700.000000MHz
clock		: 2700.000000MHz
clock		: 2700.000000MHz
clock		: 2700.000000MHz
clock		: 2700.000000MHz
clock		: 2700.000000MHz
clock		: 2700.000000MHz
clock		: 2700.000000MHz
clock		: 2700.000000MHz
clock		: 2700.000000MHz
clock		: 2700.000000MHz
clock		: 2700.000000MHz
clock		: 2700.000000MHz
```

**Additional context**
This may suggest an issue with OCC initialization. There is a difference in dmesg output, Hostboot:
```
[    1.374177] powernv-cpufreq: cpufreq pstate min 0x62 nominal 0x42 max 0x0
[    1.374179] powernv-cpufreq: Workload Optimized Frequency is enabled in the platform
[    1.375642] powernv_idle_driver registered
[    1.375945] ledtrig-cpu: registered to indicate activity on CPUs
```

coreboot:
```
[    1.353588] powernv-cpufreq: cpufreq pstate min 0x62 nominal 0x42 max 0x0
[    1.353591] powernv-cpufreq: Workload Optimized Frequency is enabled in the platform
[    1.354612] powernv-cpufreq: Frequency Control disabled from OS
[    1.354615] powernv-cpufreq: PMSR = 4242620080000000
[    1.354617] powernv-cpufreq: CPU Frequency could be throttled
[    1.355063] powernv_idle_driver registered
[    1.355437] ledtrig-cpu: registered to indicate activity on CPUs
```

**Solutions you've tried**
TBD

**Comment by tlaurion** [comment]
@krystian-hebel definitely a blocker for anyone that would be willing to use the port and any of coreboot's payload.
Any update on this @macpijan ?

**Comment by krystian-hebel** [comment]
Strange, I haven't noticed it despite stuck frequency, when we started OCC it was able to steer it. I'll check again, maybe this is regression after other changes.

@tlaurion did it work properly with Hostboot? It may have something to do with those WOF errors on your CPU.

**Comment by tlaurion** [comment]
Hostboot was starting at full speed and then fans would settle.

**Comment by tlaurion** [comment]
Will check and report back

**Comment by tlaurion** [comment]
release 0.6.1 (heads master as per Talos II merge).

Heads recovery console:
```
~ # cat /proc/cpuinfo | grep clock
clock		: 2500.000000MHz
clock		: 2500.000000MHz
clock		: 2500.000000MHz
clock		: 2500.000000MHz
clock		: 2500.000000MHz
clock		: 2500.000000MHz
clock		: 2500.000000MHz
clock		: 2500.000000MHz
clock		: 2500.000000MHz
clock		: 2500.000000MHz
clock		: 2500.000000MHz
clock		: 2500.000000MHz
clock		: 2500.000000MHz
clock		: 2500.000000MHz
clock		: 2500.000000MHz
clock		: 2500.000000MHz
```

Booted debian-11 from usb sata drive:
```
cat /proc/cpuinfo | gcat /proc/cpuinfo | grep clock
clock		: 2500.000000MHz
clock		: 2500.000000MHz
clock		: 2500.000000MHz
clock		: 2500.000000MHz
clock		: 2500.000000MHz
clock		: 2500.000000MHz
clock		: 2500.000000MHz
clock		: 2500.000000MHz
clock		: 2500.000000MHz
clock		: 2500.000000MHz
clock		: 2500.000000MHz
clock		: 2500.000000MHz
clock		: 2500.000000MHz
clock		: 2500.000000MHz
clock		: 2500.000000MHz
clock		: 2500.000000MHz
root@talos-debian:~# 
```

Unfortunately, when hostboot starts, I loose remote connection to BMC over hidden tor onion service.
Not related to this issue, but it seems that hostboot tries to renew connection in parallel and deactivates bmc ethernet trying to renew dhcp connection (on my setup this is configured for both ethernet to get dhcp address where router is giving static IP)

So I have to do something like the following to force system shutdown when I get back randomly a shell on BMC:
`while true; do ssh -o "ServerAliveInterval 5" root@server.onion "obmcutil poweroff" && echo "poweroff was successfull at $(date)" && break;done `

Otherwise the ssh connection gets disconnected randomly and getting to the shell is literally a race condition.

**Comment by krystian-hebel** [comment]
I've been doing some research, with little progress.

First of all, we aren't filling sensor IDs to OCC, so it probably isn't able to read back data like frequency or voltages. It seems that IDs of those sensors are common to all processors on Talos platforms, so they can be hardcoded, or better put into some configuration array/struct. However, applying them isn't enough to fix this issue.

During testing I noticed that at one point Skiboot reported negative (!!) core frequency: `CHIPTOD: Calculated MCBS is 0x55e63b6a (Cfreq=-1594967296 Tfreq=32000000)`. This was caused by obscured integer promotion in common coreboot code, fix was already [applied to our branch](https://github.com/Dasharo/coreboot/pull/272) and [sent upstream for review](https://review.coreboot.org/c/coreboot/+/70106). Not sure if this changed anything but this line, didn't help for the issue.

I've noticed that Hostboot played with "auto special wakeup" bits during its second OCC initialization. We've squashed those into just one, and apparently this was missed in the process. I did a quick and dirty clearing of those bits, probably at wrong point in time. I haven't seen an improvement, but I've seen a change:

```
~ # cat /proc/cpuinfo | grep clock
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2700.000000MHz
clock		: 2700.000000MHz
clock		: 2700.000000MHz
clock		: 2700.000000MHz
clock		: 2700.000000MHz
clock		: 2700.000000MHz
clock		: 2700.000000MHz
clock		: 2700.000000MHz
```

2166 MHz is safe (aka boot) frequency, 2700 MHz is nominal frequency for our CPU. I don't know yet why only half of frequencies changed, but it seems to be reliable as to which ones are slower. They also didn't change under heavier utilization, but maybe trying to stress CPU with tools available in Heads' busybox just wasn't enough. `powernv-cpufreq: Frequency Control disabled from OS` can still be found in `dmesg`, though, so I doubt it is able to change at runtime.

So far this is the only thing that changed core frequencies, so I will follow that further tomorrow.

**Comment by krystian-hebel** [comment]
Update: I've went through more than 6000 SCOM accesses done in istep 21.1 is search for differences between Hostboot and coreboot. Apart from auto special wakeup (located in quad chiplet = up to 4 cores and their common cache) mentioned earlier, I've found out that also special wakeup for single cores was asserted at all times. This may explain why all cores on a quad (first 8 threads for our CPU in listing above) had the same frequency. Next step will be to deassert those bits and see if it changes anything, for better or worse.

There were also more subtle differences that could be explained by different method of reading SCOM values - Hostboot printed them from running CPU, while for coreboot I obtained them with `pdbg` after stopping all threads.

Interestingly enough, coreboot reports Pstates enabled (but not for all cores, which is strange), while they are disabled according to Hostboot's logs...

**Comment by krystian-hebel** [comment]
CPU frequency control seems to be working with https://github.com/Dasharo/coreboot/pull/277:

```
~ # dd if=/dev/urandom of=/dev/null bs=400k count=200 & cat /proc/cpuinfo | grep clock
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
[2]-  Done                       dd if=/dev/urandom of=/dev/null bs=400k count=200
[1]-  Done                       dd if=/dev/urandom of=/dev/null bs=400k count=200
~ # dd if=/dev/urandom of=/dev/null bs=400k count=200 & cat /proc/cpuinfo | grep clock
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 3800.000000MHz
clock		: 3800.000000MHz
clock		: 3800.000000MHz
clock		: 3800.000000MHz
clock		: 3800.000000MHz
clock		: 3800.000000MHz
clock		: 3800.000000MHz
clock		: 3800.000000MHz
~ # dd if=/dev/urandom of=/dev/null bs=400k count=200 & cat /proc/cpuinfo | grep clock
clock		: 3683.000000MHz
clock		: 3683.000000MHz
clock		: 3683.000000MHz
clock		: 3683.000000MHz
clock		: 3683.000000MHz
clock		: 3683.000000MHz
clock		: 3683.000000MHz
clock		: 3683.000000MHz
clock		: 3683.000000MHz
clock		: 3683.000000MHz
clock		: 3683.000000MHz
clock		: 3683.000000MHz
clock		: 3683.000000MHz
clock		: 3683.000000MHz
clock		: 3683.000000MHz
clock		: 3683.000000MHz
~ # dd if=/dev/urandom of=/dev/null bs=400k count=200 & cat /proc/cpuinfo | grep clock
clock		: 3766.000000MHz
clock		: 3766.000000MHz
clock		: 3766.000000MHz
clock		: 3766.000000MHz
clock		: 3766.000000MHz
clock		: 3766.000000MHz
clock		: 3766.000000MHz
clock		: 3766.000000MHz
clock		: 3766.000000MHz
clock		: 3766.000000MHz
clock		: 3766.000000MHz
clock		: 3766.000000MHz
clock		: 3766.000000MHz
clock		: 3766.000000MHz
clock		: 3766.000000MHz
clock		: 3766.000000MHz
~ # dd if=/dev/urandom of=/dev/null bs=400k count=200 & cat /proc/cpuinfo | grep clock
clock		: 3800.000000MHz
clock		: 3800.000000MHz
clock		: 3800.000000MHz
clock		: 3800.000000MHz
clock		: 3800.000000MHz
clock		: 3800.000000MHz
clock		: 3800.000000MHz
clock		: 3800.000000MHz
clock		: 3800.000000MHz
clock		: 3800.000000MHz
clock		: 3800.000000MHz
clock		: 3800.000000MHz
clock		: 3800.000000MHz
clock		: 3800.000000MHz
clock		: 3800.000000MHz
clock		: 3800.000000MHz
~ # dd if=/dev/urandom of=/dev/null bs=400k count=200 & cat /proc/cpuinfo | grep clock200+0 records in
200+0 records out
81920000 bytes (78.1MB) copied, 1.380973 seconds, 56.6MB/s
200+0 records in
200+0 records out
81920000 bytes (78.1MB) copied, 1.885715 seconds, 41.4MB/s

clock		: 3766.000000MHz
clock		: 3766.000000MHz
clock		: 3766.000000MHz
clock		: 3766.000000MHz
clock		: 3766.000000MHz
clock		: 3766.000000MHz
clock		: 3766.000000MHz
clock		: 3766.000000MHz
clock		: 3766.000000MHz
clock		: 3766.000000MHz
clock		: 3766.000000MHz
clock		: 3766.000000MHz
clock		: 3766.000000MHz
clock		: 3766.000000MHz
clock		: 3766.000000MHz
clock		: 3766.000000MHz
[1]   Done                       dd if=/dev/urandom of=/dev/null bs=400k count=200
[3]   Done                       dd if=/dev/urandom of=/dev/null bs=400k count=200
~ # 200+0 records in
200+0 records out
81920000 bytes (78.1MB) copied, 1.989615 seconds, 39.3MB/s
200+0 records in
200+0 records out
81920000 bytes (78.1MB) copied, 2.875628 seconds, 27.2MB/s
200+0 records in
200+0 records out
81920000 bytes (78.1MB) copied, 1.490896 seconds, 52.4MB/s
200+0 records in
200+0 records out
81920000 bytes (78.1MB) copied, 3.412651 seconds, 22.9MB/s

[6]+  Done                       dd if=/dev/urandom of=/dev/null bs=400k count=200
[5]+  Done                       dd if=/dev/urandom of=/dev/null bs=400k count=200
[4]+  Done                       dd if=/dev/urandom of=/dev/null bs=400k count=200
[2]+  Done                       dd if=/dev/urandom of=/dev/null bs=400k count=200
~ # dd if=/dev/urandom of=/dev/null bs=400k count=200 & cat /proc/cpuinfo | grep clock
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
clock		: 2166.000000MHz
~ # 200+0 records in
200+0 records out
81920000 bytes (78.1MB) copied, 0.573054 seconds, 136.3MB/s
```

Unfortunately, this doesn't change fan speeds at all, even with all sensor data passed to OCC. Something else must be missing.

**Comment by krystian-hebel** [comment]
Fan control works with Petitboot as a payload, but not with Heads. After payload is started, host sends final IPL status that tells BMC to start working with OCC (BMC reads sensors managed by OCC and steers fans based on it). I'm not sure which part of host sends it, it may be the payload directly, or Skiboot after it receives some OPAL call from the payload.

**Comment by SergiiDmytruk** [comment]
I thought Hostboot enables BMC-to-OCC communication. `activate_occ()` in `homer.c` has this TODO:
```c
	/* TODO: Hostboot sets active sensors for all OCCs here, so BMC can start
	 * communication with OCCs. */
```

**Comment by krystian-hebel** [comment]
Hostboot starts OCC up to a point where it is able to communicate, but it doesn't inform BMC about it. BMC waits until it sees "istep 254.254" reported on IO ports 0x81/0x82 before it attempts to talk to OCC. I tried sending those from coreboot right before it jumped to Skiboot but that didn't work.

**Comment by krystian-hebel** [comment]
It seems that 254.254 is sent after Petitboot displays its menu, just before it fills HW serial numbers in top right corner, however this is based on what I've seen on two side-by-side terminals and may be impacted by SSH or serial console latency. I've found [set_ipmi_os_boot_sensor](https://git.raptorcs.com/git/talos-petitboot/tree/discover/platform-powerpc.c#n667) function in Petitboot, but nothing that directly writes to IO ports 0x81/0x82. I will check if sending that message from Heads helps, but that probably won't be today.

**Comment by krystian-hebel** [comment]
Found it: [signal_linux_start_complete](https://git.raptorcs.com/git/talos-op-build/tree/openpower/overlay/sbin/signal_linux_start_complete?id=d62c3399f5797eb08b8a15b20c012b7b7dcf1abf), called by [S17udevsettle](https://git.raptorcs.com/git/talos-op-build/tree/openpower/overlay/etc/init.d/S17udevsettle?id=d62c3399f5797eb08b8a15b20c012b7b7dcf1abf). Heads probably won't have `pnv-lpc` but `devmem` from my previous comment does the same.

Whole 0x81/0x82 debug thing was developed by RaptorCS, it isn't available in upstream repositories, or at least wasn't at the time of release. I wonder how other platforms managed to avoid it.

**Comment by krystian-hebel** [comment]
@tlaurion can you check if commands from https://github.com/Dasharo/dasharo-issues/issues/35#issuecomment-1384342830 make fans quiet? This is one thing that can't be checked remotely.

Also, how/where can it be added in Heads image so that it is run automatically on start? The order of execution with regard to other udev stuff shouldn't matter, but the sooner the better, for audible reasons.

**Comment by tlaurion** [comment]
@krystian-hebel 
![2023-01-24-095650](https://user-images.githubusercontent.com/827570/214328284-6727ba1f-1d78-4f89-8748-976070fddcae.png)
![2023-01-24-095744](https://user-images.githubusercontent.com/827570/214328426-63d4d632-06e3-4456-b9d6-0acb09ef305c.png)


Rom tested:
https://app.circleci.com/pipelines/github/osresearch/heads/506/workflows/06233d07-f99d-4d23-96c5-b74e61cc95d2/jobs/6658/artifacts

**Comment by krystian-hebel** [comment]
@tlaurion those `devmem` lines are only for the noise, did it help for that? CPU frequency will be fixed in next release, but if possible I'd like to fix both issues.

**Comment by tlaurion** [comment]
> @tlaurion those `devmem` lines are only for the noise, did it help for that? CPU frequency will be fixed in next release, but if possible I'd like to fix both issues.

@krystian-hebel : no audible difference, unfortunately.

**Comment by krystian-hebel** [comment]
Apparently it needed IPMI sensors after all. The reason why it worked in my testing was because I messed up :slightly_smiling_face: I was flashing coreboot to real flash, but Heads to file mounted by BMC, so I was using different version all the time.

https://github.com/osresearch/heads/pull/1313 should produce working set of binaries. I would suggest not merging it yet, coreboot changes are still in review so commit hash might change, but @tlaurion I'd be grateful if you can test it when you have the time.

**Comment by tlaurion** [comment]
@krystian-hebel Success!
see https://github.com/osresearch/heads/pull/1313#issuecomment-1434985130 for log and tweaks applied for faster testing (might want to update testing docs?)

**Comment by tlaurion** [comment]
> https://github.com/osresearch/heads/pull/1313 should produce working set of binaries. I would suggest not merging it yet, coreboot changes are still in review so commit hash might change, but @tlaurion I'd be grateful if you can test it when you have the time.

Waiting to be tagged for next steps on that. Will continue testing and review of docs for TPM etc

**Comment by macpijan** [comment]
Resolved as per comment: https://github.com/osresearch/heads/pull/1313#issuecomment-1435777704

**Comment by tlaurion** [comment]
It seems that we were both testing single CPU setups. I added second CPU and now the second CPU FAN is blowing at full speed. Is this same issue or should a seperate issue be opened?

**Comment by krystian-hebel** [comment]
Does the problem with CPU frequency apply to second CPU or is it fixed? Fan is a separate issue so another one should be opened. Adding more info to unrelated problem here would make it harder to find, and would mess up our internal tracking of issues.

**Comment by tlaurion** [comment]
Initial cpu frequency is fixed:
```
~ # cat /proc/cpuinfo  | grep clock
clock		: 1833.000000MHz
clock		: 1833.000000MHz
clock		: 1833.000000MHz
clock		: 1833.000000MHz
clock		: 1833.000000MHz
clock		: 1833.000000MHz
clock		: 1833.000000MHz
clock		: 1833.000000MHz
clock		: 1833.000000MHz
clock		: 1833.000000MHz
clock		: 1833.000000MHz
clock		: 1833.000000MHz
clock		: 1833.000000MHz
clock		: 1833.000000MHz
clock		: 1833.000000MHz
clock		: 1833.000000MHz
clock		: 1833.000000MHz
clock		: 1833.000000MHz
clock		: 1833.000000MHz
clock		: 1833.000000MHz
clock		: 1833.000000MHz
clock		: 1833.000000MHz
clock		: 1833.000000MHz
clock		: 1833.000000MHz
clock		: 1833.000000MHz
clock		: 1833.000000MHz
clock		: 1833.000000MHz
clock		: 1833.000000MHz
clock		: 1833.000000MHz
clock		: 1833.000000MHz
clock		: 1833.000000MHz
clock		: 1833.000000MHz

```

**Comment by tlaurion** [comment]
> Unfortunately that didn't help. What's more, even with Petitboot that call fails with `Sensor Set operation is not allowed` reported by [ipmid running on BMC](https://git.raptorcs.com/git/phosphor-host-ipmid/tree/sensorhandler.cpp?id=5d06cc6dfee3aa0c42bb181dd7e050f4f8f8dd1e#n342).
> 
> Sending 254.254 from Heads helps, although with some delay, so I wasn't able to see this initially. This can be forced from recovery shell with:
> 
> ```
> devmem 0x80060300D0010081 8 254
> devmem 0x80060300D0010082 8 254
> ```
> 
> Now all that's left is to find out which part of Skiroot + Petitboot does it, and if it can be replicated. I still don't know _where_, but at least now I know _what_ to look for...


@krystian-hebel I can open another issue, but this issue was cpu freq + fan speed.
Now cpu freq is fixated for one CPU setup, but the second CPU is still at maximum fan speed.

Heads master has board:
`export CONFIG_BOOTSCRIPT="/bin/talos-init`

where /bin/talos-init:
```
#!/bin/sh
# BMC on Talos must be informed that OS has been started in order to enable fan
# control. This is done by writing 0xFE to I/O ports 0x81 and 0x82 (in that
# order) through LPC connected to first CPU. LPC I/O space of first CPU is
# mapped to memory at 0x80060300D0010000, I/O port number has to be added to
# this address. Write can be performed using busybox's devmem applet.

devmem 0x80060300D0010081 8 254
devmem 0x80060300D0010082 8 254

# Alias for cbmem to remove need for '-f' option in every call
echo "alias cbmem='cbmem -f /sys/firmware/cbmem'" >> $HOME/.profile
export ENV=$HOME/.profile

# Proceed with standard init path
exec /bin/gui-init

```

As said in seperate discussions, it seems that we were now both testing one cpu setup. I thought that present issue being closed was consequently working for both 1cpu and 2cpu setups.
